src/goal.rs

```diff
@@ -11,192 +12,240 @@ use crate::config::*;
-    // Lookup all variables in the subst; some may be undefined if the lemma has fewer parameters
-    let target_ids_mb = self.scrutinees.iter().map(|x| subst.get(to_wildcard(x)));
-    let pairs = self
-      .scrutinees
+  fn check(&self, egraph: &Eg, _eclass: Id, subst: &Subst) -> bool {
+    // Create an iterator over triples: (variable, old canonical form, new canonical form)
+    let triples = self
+      .free_vars
       .iter()
-      .zip(target_ids_mb) // zip variables with their substitutions
-      .filter(|(_, mb)| mb.is_some()) // filter out undefined variables
-      .map(|(v, mb)| (v, extractor.find_best(*mb.unwrap()).1))
-      .collect(); // actually look up the expression by class id
-                  // Check that the expressions are smaller variables
-    let terminates = self.smaller_tuple(&pairs);
-    let premise_holds = self.check_premises(&pairs, egraph);
-    // println!("trying IH with subst {}; checks: {} {}", SmallerVar::pretty_subst(&pairs), terminates, premise_holds);
-    terminates && premise_holds
+      .map(|(x, orig_id)| {
+        let v = to_wildcard(x);
+        // Subst must have all lemma variables defined
+        // because we did the filtering when creating SmallerVars
+        let new_id = subst.get(v).unwrap();
+        // If the actual argument of the lemma is not canonical, give up
+        let new_canonical = CanonicalFormAnalysis::extract_canonical(egraph, *new_id)?;
+        // Same for the original argument:
+        // it might not be canonical if it's inconsistent, in which case there's no point applying any lemmas
+        let orig_canonical = CanonicalFormAnalysis::extract_canonical(egraph, *orig_id)?;
+        Some((x.clone(), orig_canonical, new_canonical))
+      })
+      .collect::<Option<Vec<(Symbol, Expr, Expr)>>>();
+
+    match triples {
+      None => false, // All actual arguments must be canonical in order to be comparable to the formals
+      Some(triples) => {
+        // Check that the actuals are smaller than the formals
+        // and that the actual premise holds
+        let terminates = self.smaller_tuple(&triples);
+        // Let's not check the premises if the termination check doesn't hold:
+        let sound = terminates && self.check_premises(&triples, egraph);
+        // println!("trying IH with subst {}; checks: {} {}", SmallerVar::pretty_subst(&triples), terminates, sound);
+        sound
+      }
+    }
   }
 }
 
 /// The set of constructors in an e-class.
 /// The order of variants is important: since we use the derived order during the merge.
 #[derive(Debug, PartialEq, PartialOrd, Eq, Ord, Clone)]
-pub enum Constructors {
-  /// No constructors
-  Zero,
-  /// Single constructor
-  One(SymbolLang),
-  /// At least two different constructors (inconsistency)
-  Two(SymbolLang, SymbolLang),
+pub enum CanonicalForm {
+  /// This class has neither constructors nor variables
+  Stuck,
+  /// This class has a variable but no constructors
+  Var(SymbolLang),
```

Comment:
You're right that it would be sufficient for a variable to just store the `Symbol`, but for the sake of symmetry with the other cases, I think it's okay to store the enode too.